zk/stages/stage_sequence_execute.go

```diff
@@ -47,6 +44,83 @@ func SpawnSequencingStage(
 		defer tx.Rollback()
 	}
 
+	lastBatch, err := stages.GetStageProgress(tx, stages.HighestSeenBatchNumber)
+	if err != nil {
+		return err
+	}
+
+	highestBatchInDS, err := cfg.datastreamServer.GetHighestBatchNumber()
+	if err != nil {
+		return err
+	}
+
+	if lastBatch >= highestBatchInDS {
+		err = sequencingStageStep(s, u, tx, ctx, cfg, quiet, nil)
+		if err != nil {
+			return err
+		}
+	} else {
+		log.Info(fmt.Sprintf("[%s] Last batch %d is lower than highest batch in datastream %d, resequencing...", s.LogPrefix(), lastBatch, highestBatchInDS))
+
+		latestForkId, err := stages.GetStageProgress(tx, stages.ForkId)
+		if err != nil {
+			return err
+		}
+		dsClient := client.NewClient(ctx, fmt.Sprintf("%s:%d", cfg.zk.DataStreamHost, cfg.zk.DataStreamPort), cfg.zk.DatastreamVersion, cfg.zk.L2DataStreamerTimeout, uint16(latestForkId))
```

Comment:
I am not sure if the data stream client is initiated in the sequencer, because the sequencer in most cases will only have the data stream server. I think it is okay to only create the client when it is needed when resequencing.